Changelog CLI: Add command to make entry from PR description

psealock opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/38125

Add a CLI utility to add a changelog entry from details supplied in a PR description.

Note on validation: I'm not doing any validation of the input. the script will fail at the Jetpack Changelogger creation step if input not valid.

Note on changelog of dependencies: In a follow up I'd like to address changelogs being added to dependencies based on dependency graphs. If a dependency gets updated, it will get an entry and so will all projects consuming it.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites
1. You must have a `GITHUB_TOKEN` environment variable available in your PATH. This needs to be set to a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with repo permissions enabled.

#### Setup your fork
1. Fork this repo.
2. Push this branch to your fork (`add/changelog-determine-projects`).
3. Create a branch off this one called `test/add/changelog-determine-projects`.
4. In the new branch, create a few changes to the codebase. Be sure to include some changes to projects with changelogger like `woocommerce` and some to projects that do not, like any of our `tools/`.
5. Push `test/add/changelog-determine-projects` to your fork using `--no-verify` to skip changelog checks.
6. Create a pull request on your fork to merge `test/add/changelog-determine-projects` into `add/changelog-determine-projects`.
7. Note the pull request number

#### Test the command
1. Switch back to this PR's branch, `add/changelog-determine-projects` and run the command. The PR doesn't have the corresponding description check, so nothing will happen. Confirm the message makes sense.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
2. Add this template to the bottom of the open PR description and fill in the relevant details. Confirm the data entry aspect is ok.
```
### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

- [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance
<!-- Choose only one -->
- [ ] Patch
- [ ] Minor
- [ ] Major

#### Type
<!-- Choose only one -->
- [ ] Fix - Fixes an existing bug
- [ ] Add - Adds functionality
- [ ] Update - Update existing functionality
- [ ] Dev - Development related task
- [ ] Tweak - A minor adjustment to the codebase
- [ ] Performance - Address performance issues
- [ ] Enhancement

#### Message
<!-- Add a changelog message here -->

#### Comment
<!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>
```
3. Now run the command again. This time it should succeed in creating changelog entries and adding directly to the pull request's branch. Confirm changelogs were added only to public projects that have Jetpack Changelogger.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
4. In the `test/add/changelog-determine-projects` branch undo the change to `woocommerce`, if applicable. If not, undo any of the other changes. Push changes to your fork and re run the command from `add/changelog-determine-projects` to see the changelog for the project that no longer has changes go away.
```
pnpm run utils changefile <prNumber> -o <forkOwner>
```
6. Finally make a change in `test/add/changelog-determine-projects` to woocommerce-admin. Push and rerun the command. A changelog entry to `woocommerce` should be added.

<!-- End testing instructions -->

psealock committed:
get extract working

psealock committed:
regexes

psealock committed:
log data

psealock committed:
setOutput

psealock committed:
TS repo functions

psealock committed:
doc lib

psealock committed:
logger output regardless

psealock committed:
reorganize for single command

psealock committed:
add temprepo

psealock committed:
get all changelogger projects

psealock committed:
get touched projects

psealock committed:
getTouchedProjectsRequiringChangelog

psealock committed:
save

psealock committed:
mappedTouchedProjects

psealock committed:
add changelog files

psealock committed:
run changelogger

psealock committed:
get the whole thing working

psealock committed:
clean up main file

psealock committed:
make functions testable

psealock committed:
scaffold testing suite

psealock committed:
test getAllProjects

psealock committed:
test getChangeloggerProjects

psealock committed:
test intersectTouchedFilesWithChangeloggerProjects

psealock committed:
jsDoc

psealock committed:
save

psealock committed:
save

psealock committed:
iron out kinks?

psealock committed:
whitespace

psealock committed:
install

psealock committed:
remove all the changelogs in every project

psealock committed:
simplify project paths

psealock committed:
jsDoc

psealock committed:
get details after check

psealock committed:
replace all

psealock committed:
use owner/name to compare for community contribs

psealock committed:
don't count changelogs as touched files

psealock committed:
typo

psealock committed:
remove console

lsinger commented on tools/monorepo-utils/src/core/github/types.ts:
```diff
@@ -3,5 +3,8 @@
+export type CreatePullRequestEndpointResponse =
 	Endpoints[ 'POST /repos/{owner}/{repo}/pulls' ][ 'response' ];
+
+export type PullRequestEndpointResponse =
```
If we call the creating one `Create...`, maybe we should call the fetching one `Get...` or `Read`.

psealock replied:
Good call, fixed in https://github.com/woocommerce/woocommerce/pull/38357/commits/5b684f89e59f1131c82579d032a94957982fc4e1

lsinger commented on tools/monorepo-utils/src/changelog/index.ts:
```diff
@@ -0,0 +1,199 @@
+	getTouchedProjectsRequiringChangelog,
+} from './lib/projects';
+
+const program = new Command( 'changelog' )
```
For now it's focused on automating changlog creation from pull request details, so something more specific might be easier to understand:

```suggestion
const program = new Command( 'changelog-from-pr' )
```

Having to type more won't matter as much as it'll usually run from a CI context.

jonathansadowski replied:
In general, I think we misuse the term "changelog" 😄.

* "changelog" / "changelog file": the compiled list of all changes in a release.
* "change file": a file containing info about a single change tied to a PR.

The latter is what we're generating here... change files, not changelog files.

psealock replied:
How about `pnpm utils changelog file 8988` as a command?

lsinger replied:
`pnpm utils changefile 8988`?

lsinger commented on tools/monorepo-utils/src/changelog/README.md:
```diff
@@ -0,0 +1,3 @@
+# Changelog CLI Utility
+
+CLI for manipulating pull request changelog entries
```
```suggestion
CLI for creating changelog entries from pull request details. 

Usage: `pnpm utils changelog <pr-number>`

lsinger commented on tools/monorepo-utils/src/changelog/index.ts:
```diff
@@ -0,0 +1,199 @@
+} from './lib/projects';
+
+const program = new Command( 'changelog' )
+	.description( 'Changelog utilities' )
```
```suggestion
	.description( 'Changelog generation from pull request details' )
```

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
 };
+
+/**
+ * Get a pull request from Github.
```
```suggestion
 * Get a pull request from GitHub.
```

¯\_(ツ)_/¯

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
+};
+
+/**
+ * Determine if a pull request is coming from a community contribution. ie, not from a member of the woocommerce org.
```
```suggestion
 * Determine if a pull request is coming from a community contribution, i.e., not from a member of the WooCommerce organization.
```

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
+) => {
+	return (
+		pullRequestData.author_association === 'CONTRIBUTOR' ||
+		// Its possible a MEMBER can open a PR from thier own fork.
```
```suggestion
		// It's possible a MEMBER can open a PR from their own fork.
```

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
+ * @return {boolean} if a pull request is coming from a community contribution.
+ */
+export const isCommunityPullRequest = (
+	pullRequestData: PullRequestEndpointResponse[ 'data' ],
```
As this function uses only two properties from the endpoint response -- `author_association` and the repo's `full_name` -- I wonder if it would be cleaner to really only have those two be part of the function's interface.

psealock replied:
You have a good point. The type would look something like this:
```
{
  author_association: string,
  head: {
    repo: {
      full_name: string,
    }
  }
}
```
For me thats inferring a lot of assumptions about the shape of the data. I'd prefer to use the one from Octokit because its less error prone.

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
+ * @return {Promise<object>}     pull request data.
+ */
+export const getPullRequest = async (
+	options: { owner: string; name: string },
```
Not really a review comment but rather a question: `isCommunityPullRequest` has `owner` and `name` as proper arguments, while here they're part of the `options` argument (however, not so `prNumber`). 

What's the intention for these different style's, and what's the heuristic when to choose what?

psealock replied:
oops, yes they can all be in `options`. Fixed in https://github.com/woocommerce/woocommerce/pull/38357/commits/8bde8b6525236cbf4378f1ffef13d76aebdd6421

lsinger commented on tools/monorepo-utils/src/core/github/repo.ts:
```diff
@@ -167,3 +170,51 @@ export const createPullRequest = async ( options: {
+		}
+	);
+
+	//@ts-ignore
```
This confused me as I couldn't see any 'why' for this. I then found this comment in a different function. Is this the rationale? 

```suggestion
	//@ts-ignore There is a type mismatch between the graphql schema and the response. pullRequest.data.head.repo.has_discussions is a boolean, but the graphql schema doesn't have that field.
```

When I removed the `@ts-ignore` in VS Code it did not complain. How / when would I be able to trigger it complaining -- only when building TS as part of a build step? Or is my IDE setup possibly incomplete here?

psealock replied:
When I remove it in my IDE, I get the following which is different than your suggestion above.

<img width="829" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/9622ac72-901e-4a8a-84ec-3b6430cacba8">

I have made sure I'm on all the latest Octokit packages and this still persists. I don't know why there is an error, so I didn't add an explanation. Open to suggestions here

lsinger replied:
Weird and unfortunate, but nothing we can or should solve as part of this PR. 

I'd add a comment giving maybe the last sentence of the error message and basically something like "we don't know why this error comes up, we'll try to solve it in a future iteration" -- just so that someone not super familiar with the code (e.g. us in six months) will understand our why.

lsinger commented on tools/monorepo-utils/src/changelog/index.ts:
```diff
@@ -0,0 +1,199 @@
+			}
+
+			Logger.notice( `Checking out remote branch ${ branch }` );
+			await checkoutRemoteBranch( tmpRepoPath, branch );
```
It seems we only need the PR branch here, yet we clone the whole repo. We also don't need the whole commit history. Could make sense her to optimize a bit? 

I timed this once locally (so take with a grain of salt) and got the following times:

`git clone git@github.com:woocommerce/woocommerce.git && cd woocommerce && git checkout add/changelog-determine-projects`: 48 seconds

`git clone -b add/changelog-determine-projects --single-branch git@github.com:woocommerce/woocommerce.git`: 33 seconds

`git clone -b add/changelog-determine-projects --single-branch --depth=1 git@github.com:woocommerce/woocommerce.git`: 10 seconds

psealock replied:
Good call here! I think the best optimization will be to the `cloneRepo` and `cloneRepoShallow` functions we already have. I can add the branch for `--single-branch` as argument to these functions and have the same effect as your third scenario.

I think I'll tackle this in a follow up.

lsinger commented on tools/monorepo-utils/src/changelog/index.ts:
```diff
@@ -0,0 +1,199 @@
+					execSync( cmd, { cwd: tmpRepoPath, stdio: 'inherit' } );
+				} );
+			} catch ( e ) {
+				Logger.error( e );
```
Shouldn't we exit early here if the above went wrong?

psealock replied:
`Logger.error` does exit with an error, so we're good here.

lsinger commented on tools/monorepo-utils/src/changelog/lib/__tests__/projects.ts:
```diff
@@ -0,0 +1,140 @@
+
+		expectedChangeLoggerProjects.forEach(
+			( expectedChangeLoggerProject ) => {
+				expect( projects ).toContain( expectedChangeLoggerProject );
```
Shouldn't `projects` be `changeloggerProjects` here, and wouldn't `projects` be unnecessary here then?

psealock replied:
Yikes, nice catch! Fixed in https://github.com/woocommerce/woocommerce/pull/38357/commits/799c80914b66d124ba28c5622773458d6762704a

lsinger commented on tools/monorepo-utils/src/changelog/lib/__tests__/projects.ts:
```diff
@@ -0,0 +1,140 @@
+		expect( intersectedProjects ).toContain( 'projects/very-cool-project' );
+	} );
+
+	it( 'getTouchedChangeloggerProjectsPathsMappedToProjects should map plugins and js packages to the correct name', async () => {
```
In my personal nitpick mode this doesn't check whether paths are mapped to the correct project, as we're checking across four different projects here. Could be nice to add checks that ensure a direct mapping -- e.g., with files only touching a single project. (Say, `touchedWooCommerceFiles` touch only the `plugins/woocommerce` project, and `touchedBetaTesterFiles` only touch the `plugins/beta-tester` project.)

lsinger commented on tools/monorepo-utils/src/changelog/lib/__tests__/projects.ts:
```diff
@@ -0,0 +1,140 @@
+
+	it( 'getTouchedChangeloggerProjectsPathsMappedToProjects should handle woocommerce-admin projects mapped to woocommerce core', async () => {
+		const touchedFiles = [
+			'plugins/beta-tester/src/index.js',
```
I feel like the inclusion of `beta-tester` distracts a bit from the purpose of this test and could be removed, especially since that's covered by the prior test.

psealock replied:
fixed

lsinger commented on tools/monorepo-utils/src/changelog/lib/github.ts:
```diff
@@ -0,0 +1,143 @@
+/**
```
I really, really like that you're including tests for `projects.ts`. Would be nice to have some for `github.ts` as well, especially seeing how it's heavily relying on regular expressions. Always nice to have those tested well, imho. 

This could also help enforce some more separation between the fetching and the parsing tasks.

psealock replied:
yes, good call. I added tests and reworked the function to have an improved regex

lsinger replied:
Thanks so much!

lsinger commented on tools/monorepo-utils/src/changelog/lib/github.ts:
```diff
@@ -0,0 +1,143 @@
+ * @param {string} body pull request description.
+ * @return {boolean} if the pull request description activates the changelog automation.
+ */
+export const getShouldAutomateChangelog = ( body: string ) => {
```
Not sure whether / where we have naming conventions documented, but for boolean getters I've usually seen (in other projects) the `get` omitted to then have it start with something like the `should` here, or `will` or `does` or `is` or something similar.

psealock replied:
yeah what a conundrum! I went back and forth on this one a few times and settled on this so that I could have the const appropriately named in the code:

```js
const shouldAutomateChangelog = getShouldAutomateChangelog( body );
``` 
If the function is called `shouldAutomateChangelog`, what do I call the const? Maybe `willAutomateChangelog` for the function?

lsinger replied:
Hmmm, when taking a principled vantage point I'd say that it shouldn't be the responsibility of a module how its callers will cache its results. 

Pragmatically speaking, I'd probably not even assign it to a const at all in index.ts, as it's used only once. 

Another pragmatic stance could be to name the const `_shouldAutomateChangelog`, even though that has implications of class membership in other contexts. 

All in all I'd say you've taken a pragmatic approach as well that is completely fine, and I'm happy to have things move forward with this or one of the other mentioned solutions. ☺️

lsinger commented on tools/monorepo-utils/src/changelog/lib/github.ts:
```diff
@@ -0,0 +1,143 @@
+ * @return {void|string} changelog significance.
+ */
+export const getChangelogSignificance = ( body: string ) => {
+	const patchRegex = /\[x\] (Patch)\r\n/gms;
```
Shouldn't we be able to combine those three expressions into one?

psealock replied:
Yes, for sure. Do you think the tradeoff on readability is worth the reduction in lines of code?

psealock replied:
ok, I did this. Should be smaller and readable

```js
const regex = /\[x\] (Patch|Minor|Major)\r\n/gm;
```

lsinger commented on tools/monorepo-utils/src/changelog/lib/github.ts:
```diff
@@ -0,0 +1,143 @@
+ * @return {void|string} changelog type.
+ */
+export const getChangelogType = ( body: string ) => {
+	const fixRegex = /\[x\] (Fix) -/gm;
```
Same as above -- quite some repetition here with the prefix and postfix, can't we combine those into a single expression?

psealock replied:
same treatment here

lsinger reviewed (commented):
I got assigned to review #38327 and hadn't submitted my comments yet when that PR was closed, so I've moved over my comments from there and added new ones for the additional changes. 

I'm submitting them not as requests, but out of curiosity and in an attempt to learn and grow more accustomed with Atlas's code base. Reading your code was helpful and interesting.

jonathansadowski reviewed (commented):
Nice work on this PR. I have some questions about how this will function:

* You mentioned that you didn't do any validation on the PR description. Would we add that before merging this? Would it validate that there is either a valid PR description or a valid change file for each affected project?
* When / how would this be run? It appears to update the base branch of the PR automatically. Will this be a problem with `trunk` (the base of most PRs) being a protected branch?
* If this does add the change files directly to `trunk`, will this cause problems with the code that detects and automatically adds the PR number?

psealock committed:
better GH PR type name

psealock committed:
Update tools/monorepo-utils/src/changelog/README.md

Co-authored-by: Leif Singer <leif@automattic.com>

psealock committed:
Update tools/monorepo-utils/src/core/github/repo.ts

Co-authored-by: Leif Singer <leif@automattic.com>

psealock committed:
Update tools/monorepo-utils/src/core/github/repo.ts

Co-authored-by: Leif Singer <leif@automattic.com>

psealock committed:
Update tools/monorepo-utils/src/core/github/repo.ts

Co-authored-by: Leif Singer <leif@automattic.com>

psealock committed:
consolidate args in getPullRequest

psealock committed:
Merge branch 'add/changelog-determine-projects' of github.com:woocommerce/woocommerce into add/changelog-determine-projects

psealock committed:
fix changeloggerProjects tests

psealock commented:
@jonathansadowski, thanks for taking a look

>You mentioned that you didn't do any validation on the PR description. Would we add that before merging this? Would it validate that there is either a valid PR description or a valid change file for each affected project?

Jetpack changelogger fails at creation of the change file when input is incorrect and that error gets surfaced, so I'm not too concerned about that. But validation that may be worthwhile is when someone checks multiple boxes, ie `patch` and `minor`. Currently, my regex would choose the first one and continue on. 

Now that I think about it, yeah probably a good idea to add that in to this PR.

>When / how would this be run? It appears to update the base branch of the PR automatically. Will this be a problem with trunk (the base of most PRs) being a protected branch? 

It doesn't update the base branch, it updates the head branch. So the additional changefiles get created and added to the pull request itself. I'm intending to run this on PR creation and every update to the PR description.

>If this does add the change files directly to trunk, will this cause problems with the code that detects and automatically adds the PR number?

Again, won't be an issue because the changefiles will get added to the PR as if the author had created them, then when the PR is merged they'll get added to trunk and dealt with normally.

psealock committed:
getChangelogSignificance reworked and tested

psealock committed:
test getChangelogType

psealock committed:
test getChangelogDetails

psealock commented:
@lsinger @jonathansadowski I have added most of your suggestions, reworked the regexes, and added tests to the functions interpreting github PR descriptions, including validation for common issues like selecting multiple types, etc.

Could you give this another look?

lsinger commented:
> Jetpack changelogger fails at creation of the change file when input is incorrect and that error gets surfaced, so I'm not too concerned about that. But validation that may be worthwhile is when someone checks multiple boxes, ie `patch` and `minor`. Currently, my regex would choose the first one and continue on.
> 
> Now that I think about it, yeah probably a good idea to add that in to this PR.

Maybe adding a sort of feature flag to this to start with could be useful? Something like ... only run on PRs with label `ff: changefile generation`.

lsinger commented on tools/monorepo-utils/src/changelog/lib/__tests__/github.ts:
```diff
@@ -0,0 +1,395 @@
+			'\r\n' +
+			'</details>';
+
+		getChangelogSignificance( body );
```
I'd also assign this to a const and check the result, even if the expected result is null or undefined in this case. But by making it explicit in the test we're also documenting how callers might guard against invalid inputs. 

We could also be throwing an exception here or be returning a special error value. By looking at the test, the reader would know without having to find this exact path in the actual code.

lsinger commented on tools/monorepo-utils/src/changelog/lib/github.ts:
```diff
@@ -0,0 +1,168 @@
+		Logger.error(
+			'Only patch changes can have a comment. Please change the significance to patch or remove the comment'
+		);
+		// Logger.error has a process.exit( 1 ) call, this return is purely for testing purposes.
```
Can these (a few instances throughout the file) be removed now or will we need them in further testing post-merge?

psealock replied:
They are needed for unit tests, so will need to stay

lsinger commented:
> @lsinger @jonathansadowski I have added most of your suggestions, reworked the regexes, and added tests to the functions interpreting github PR descriptions, including validation for common issues like selecting multiple types, etc.
> 
> Could you give this another look?

Thank you @psealock for making all those changes! From my perspective there aren't any blockers now, but I'll leave the final approval to @jonathansadowski. 

I do wonder whether we should rework some things in a follow-up PR to address the terminology discussion (change vs. change fit vs. change log).

psealock committed:
changefile

psealock committed:
feedback

psealock committed:
test undefined

psealock commented:
@lsinger 

* I changed the command to `changefile`
* Made sure to test undefined as the output of `getChangelogType` and other when errors occur
* I renamed `shouldAutomateChangelog`
* Added a comment to the ts-ignore

jonathansadowski commented:
Hmm, this worked the first pass around, but I just retested it now, and I'm getting this...

```
pnpm run utils changefile 1 -o jonathansadowski

> woocommerce-monorepo@ utils /Users/jonathan/woo/woocommerce
> node ./tools/monorepo-utils/dist/index.js "changefile" "1" "-o" "jonathansadowski"

 __        __           ____
 \ \      / /__   ___  / ___|___  _ __ ___  _ __ ___   ___ _ __ ___ ___
  \ \ /\ / / _ \ / _ \| |   / _ \| '_ ` _ \| '_ ` _ \ / _ \ '__/ __/ _ \
   \ V  V / (_) | (_) | |__| (_) | | | | | | | | | | |  __/ | | (_|  __/
   _\_/\_/ \___/ \___/ \____\___/|_| |_| |_|_| |_| |_|\___|_|  \___\___|
  | | | | |_(_) |___
  | | | | __| | / __|
  | |_| | |_| | \__ \
   \___/ \__|_|_|___/

✔ Getting pull request data for PR number 1... complete.
No changelog message found
```

It looks like [the PR](https://github.com/jonathansadowski/woocommerce/pull/1) has a correctly formatted changelog entry in the message body though

psealock committed:
regex to account for deleted comment

jonathansadowski reviewed (approved):
Nice work. This is pretty robust, and love all the added tests! Thanks for updating the regex to account for the comment being removed. 👍
